platforms/core-runtime/service-registry-impl/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -786,23 +801,32 @@ protected void bind() {
         }
 
         @Override
-        public Service getService(Type serviceType) {
+        public Service getService(Type serviceType, @Nullable ServiceAccessToken token) {
+            if (!accessScope.contains(token)) {
+                return null;
+            }
```

Comment:
ℹ️  This is the main change to the lookup algorithm.

Each lookup is now happening with an access token. The token is null for lookups originating from outside the service registry. In all other cases, the token comes from a service provider that tries to resolve its dependencies.

Then, the scope of the specific service determines whether it can be returned to the requester. There are currently only two scopes: public (anyone can access with any token) and private (can be accessed only with the same token).